### PR TITLE
Changeable npc currency

### PR DESCRIPTION
**Pull request by dex-89**
image

#2117 **In global tibia Cledwyn work: hi -> token(s) -> silver -> trade window and sell for silver tokens**

Thanks for @FakeShinoda

### CODE REVIEW

File: data/npc/scripts/cledwyn.lua [hunk 1]
```diff
@@ -1,71 +1,91 @@
...
+			if amount > 1 then
+				currencyName = ItemType(Npc():getCurrency()):getPluralName():lower()
+			else
+				currencyName = ItemType(Npc():getCurrency()):getName():lower()
+			end
+			player:addItem(itemsTable[item].id, amount)
+			return player:sendTextMessage(MESSAGE_INFO_DESCR,
+						"Bought "..amount.."x "..itemsTable[item].name.." for "..itemsTable[item].buy * amount.." "..currencyName..".")
+		else
+			return player:sendTextMessage(MESSAGE_INFO_DESCR, "You don't have enough "..currencyName..".")
```
Comment (FakeShinoda) [inline data/npc/scripts/cledwyn.lua:69]: You do not need to specify whether or not you have the currency, the purchase option will only be available if you have the currency in stock.

File: data/npc/scripts/cledwyn.lua [hunk 2]
```diff
@@ -78,68 +98,65 @@ local voices = {
...
+		if not chargeItem[charge] then
+			return false
+		else
+			if (player:getItemCount(25172) >= 2) and (player:getItemCount(chargeItem[charge].noChargeID) >= 1) then
+				player:removeItem(25172, 2)
+				player:removeItem(chargeItem[charge].noChargeID, 1)
+				local itemAdd = player:addItem(chargeItem[charge].ChargeID, 1)
+				npcHandler:say("Ah, excellent. Here is your " .. itemAdd:getName():lower() .. ".", cid)
+			else
+				npcHandler:say("Sorry, friend, but one good turn deserves another. Bring enough tokens and it's a deal.", cid)
```
Comment (FakeShinoda) [inline data/npc/scripts/cledwyn.lua]: Same as above

File: src/npc.cpp [hunk 3]
```diff
@@ -147,6 +147,17 @@ bool Npc::loadFromXml()
...
 	floorChange = npcNode.attribute("floorchange").as_bool();
 
+	pugi::xml_attribute coin;
+	if ((coin = npcNode.attribute("currency"))) {
+        const ItemType& it = Item::items[pugi::cast<uint16_t>(coin.value())];
+		currencyServerId = it.id;
+        currencyClientId = it.clientId;
+	} else {
+        const ItemType& it = Item::items[ITEM_GOLD_COIN];
+		currencyServerId = it.id;
```
Comment (FakeShinoda) [inline src/npc.cpp:157]: Fix indentation

File: src/npc.cpp [hunk 4]
```diff
@@ -147,6 +147,17 @@ bool Npc::loadFromXml()
 	attackable = npcNode.attribute("attackable").as_bool();
 	floorChange = npcNode.attribute("floorchange").as_bool();
 
+	pugi::xml_attribute coin;
+	if ((coin = npcNode.attribute("currency"))) {
+        const ItemType& it = Item::items[pugi::cast<uint16_t>(coin.value())];
+		currencyServerId = it.id;
```
Comment (FakeShinoda) [inline src/npc.cpp:153]: Fix indentation

**Review by FakeShinoda — CHANGES_REQUESTED** [review]

File: data/npc/scripts/cledwyn.lua (hunk 1 above)
Comment (dex-89) [reply data/npc/scripts/cledwyn.lua:69]: if amount > 1 then
     currencyName = ItemType(Npc():getCurrency()):getPluralName():lower()
else
     currencyName = ItemType(Npc():getCurrency()):getName():lower()
 end

plural name... 1 silver token, 2,3,... silver token**s**

File: data/npc/scripts/cledwyn.lua (hunk 2 above)
Comment (dex-89) [reply data/npc/scripts/cledwyn.lua]: I only fixed the indentation here

**Review by FakeShinoda — APPROVED** [review]

**Review by majestyotbr — APPROVED** [review]

**Review by majestyotbr — APPROVED** [review]

**Review by majestyotbr — APPROVED** [review]

**Comment by majestyotbr** [comment]
Tested again, i found a small issue when i say `trade`, the npcs answer is: `Sorry, I'm not offering anything.`

Cledwyn should not answer anything for `trade.`
Yana should answer: `I have creature products for the imbuements strike, vampirism and void. Make your choice, please!`

**Comment by dex-89** [comment]
> Tested again, i found a small issue when i say `trade`, the npcs answer is: `Sorry, I'm not offering anything.`
> 
> Cledwyn should not answer anything for `trade.`
> Yana should answer: `I have creature products for the imbuements strike, vampirism and void. Make your choice, please!`

It was after merge PR #2231. Fixed

**Review by majestyotbr — APPROVED** [review]

**Review by hyresu — APPROVED** [review]

**Comment by dex-89** [comment]
> I found a bug
> 
> Missing to update "Stock" when buy/sell
> 
> https://prnt.sc/wgd5ju

Thank you for your information. I'll check it today and try fix mistake.

**Pull request merged**